tools/python/boutiques/bosh.py

```diff
@@ -277,6 +281,37 @@ def evaluate(*params):
     return query_results[0] if len(query_results) == 1 else query_results
 
 
+def test(*params):
+
+    parser = ArgumentParser("Perform all the tests defined within the given descriptor")
+    parser.add_argument("descriptor", action="store", help="The Boutiques descriptor.")
+    result = parser.parse_args(params)
+
+    # Generation of the invocation schema (and descriptor validation).
+    invocation(result.descriptor)
+    
+    # Extraction of all the invocations defined for the test-cases.
+    descriptor = json.loads(open(result.descriptor).read())
+    
+    if (not descriptor.get("tests")):
+        raise SystemExit("No test found in descriptor")
+    
+    for test in descriptor["tests"]:
+        # Create temporary file for the invocation() function.
+        invocation_JSON = test["invocation"];
+        temp_invocation_JSON = tempfile.NamedTemporaryFile(suffix=".json", delete=False)
+        temp_invocation_JSON.write(json.dumps(invocation_JSON).encode())
+        temp_invocation_JSON.seek(0)
```

Comment:
Yes that's something that I have highlighted in the description above.

Modifying bosh so that at least bosh invocation can take objects as invocations would make the code in both bosh test and conftest.py much cleaner. It would also remove the need to rely on intermediate files just like you have stated.

So yes I agree :) 

I think for all the bosh functions, it is simply a matter of adding an if case that checks if the invocation/descriptor parameter is a variable of type dict. In which case bosh will simply skip over the part about opening and loading up the JSON.

Reply:
Rather than writing it each time, let's write a little utility and call it in our functions?

Reply:
I agree, but can we merge this first and then discuss how the API should be modified to support objeects?

Reply:
I'm happy merging first, but I don't think the API will change at all - each routine will just evaluate the object it receives (dict or file), and will return the dict version of that data.

Reply:
okay then feel free to do it if it's clear in your mind @Trajan64

Reply:
Yes sure :) The modifications should be very light.
And I agree with @gkiar's perspective, I'll try to implement a small function to perform the task.

Reply:
Turns out there's no clean way to handle this issue.
It is apparently impossible to allow `argparse` to accept dictionary objects as arguments. 
Which makes sense, but I wanted to enable the usage of the inner bosh functions directly with dictionary objects.
And as `argparse` is included in the beginning of every functions, this is just impossible to do.

There are multiple alternatives that can be used as work-around, but none of them are clean or simple:
- (1) Allow representation of dictionaries as raw `str` objects. This is tricky as each functions will have to differentiate between a string representing a dictionary, and a string representing a file path.
  - One way to perform this differentiation is to check if the first and last characters of the `str` object are respectively `{` and `}`
  - Another is to setup a try except block with a `json.loads` (loading for raw data) and if it works, we can be sure its a proper dictionary represented as a string and not a file path.
- (2) I think I can bypass `argparse` behavior by setting up a list containing of a string as first element, and the proper JSON dictionary as second element
- (3) Put all the `argparse` logic in the bosh function.

Also note that the modifications required go beyond `bosh.py`: `localExec.py` (`__init__` and `readInput` functions) and `validate.py` (`__init__` only) need to be modified so that they can choose to open a file with the provided object as string path, or read directly from the object to get the JSON.

For now I'll leave the problem as it is and will await your directions on the matter.

Reply:
@Trajan64 no problem - you don't need to implement this in this PR. I have done something similar to this before, so will take a crack at it myself later on. I suspect your assessment is correct, but that there are simplification we can make to still have this work for us (i.e. wrap it as a string as you suggest, etc.)..

I mentioned to @glatard that I've made plans to overhaul the API organization a bit again with some recent lessons learned, and will do this then. So as @glatard suggested upstream here, please ignore this suggestion in the PR, and let us know when all other suggested changes are done.

Cheers!